File: libs/ontology-kernel/src/models/submission_criteria.rs
//! TASK C — Submission criteria AST.
//!
//! Authored representation of the per-action gate documented in
//! `docs_original_palantir_foundry/foundry-docs/Ontology building/Define
//! Ontologies/Action types/Submission criteria.md`.
//!
//! The AST is persisted as JSONB on `action_types.submission_criteria`
//! (migration `20260502000000_add_submission_criteria.sql`). The
//! evaluator lives in [`crate::domain::submission_eval`].
//!
//! Each node owns its own optional `failure_message`; per the Foundry
//! contract, when a node fails the evaluator surfaces ITS message and
//! suppresses the messages of failing descendants. Leaves without an
//! explicit message contribute a synthesized "criterion failed" string so
//! the user-facing list is never empty.

use serde::{Deserialize, Serialize};
use serde_json::Value;

/// Operand resolution kinds available in the authoring UI. Mirrors the
/// Foundry "Current User" / "Parameter" / "Property of parameter" / "Static"
/// templates.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum Operand {
    /// Reference to an action input field by name. Resolves to the
    /// already-coerced parameter value.
    Param { name: String },
    /// Property dereference of a parameter value: `param.property`. The
    /// referenced parameter must be an object map (e.g. an object reference
    /// expanded via `materialize_parameters`).
    ParamProperty { param: String, property: String },
    /// Attribute of the calling user, sourced from the JWT claims.
    CurrentUser { attribute: UserAttr },
    /// Static value authored at design time. Always JSON.
    Static { value: Value },
}

/// Subset of `auth_middleware::Claims` exposed to submission criteria.
/// Anything outside this set must NOT be addressable from authored criteria
/// to keep the evaluation surface auditable and deterministic.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum UserAttr {
    UserId,
    Email,
    OrganizationId,
    Roles,
    Permissions,
    AuthMethods,
}

/// Operators supported by the evaluator. The vocabulary matches
/// `Action types/Submission criteria.md` exactly (14 ops).
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum Operator {
    Is,
    IsNot,
    Matches,
    Lt,
    Lte,
    Gt,
    Gte,
    Includes,
    IncludesAny,
    IsIncludedIn,
    EachIs,
    EachIsNot,
    IsEmpty,
    IsNotEmpty,
}

/// Tree of submission criteria evaluated in
/// [`crate::domain::submission_eval::evaluate`].
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SubmissionNode {
    Leaf {
        left: Operand,
        op: Operator,
        right: Operand,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        failure_message: Option<String>,
    },
    All {
        children: Vec<SubmissionNode>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        failure_message: Option<String>,
    },
    Any {
        children: Vec<SubmissionNode>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        failure_message: Option<String>,
    },
    Not {
        child: Box<SubmissionNode>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        failure_message: Option<String>,
    },
}

impl SubmissionNode {
    /// Convenience constructor used by tests / programmatic seeding.
    pub fn leaf(left: Operand, op: Operator, right: Operand) -> Self {
        Self::Leaf {
            left,
            op,
            right,
            failure_message: None,
        }
    }

    /// Returns the failure message attached to this node, if any.
    pub fn failure_message(&self) -> Option<&str> {
        match self {
            Self::Leaf {
                failure_message, ..
            }
            | Self::All {
                failure_message, ..
            }
            | Self::Any {
                failure_message, ..
            }
            | Self::Not {
                failure_message, ..
            } => failure_message.as_deref(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn round_trip_serde_preserves_shape() {
        let node = SubmissionNode::All {
            failure_message: Some("policy".to_string()),
            children: vec![
                SubmissionNode::Leaf {
                    left: Operand::CurrentUser {
                        attribute: UserAttr::Roles,
                    },
                    op: Operator::Includes,
                    right: Operand::Static {
                        value: json!("ontology.editor"),
                    },
                    failure_message: Some("must be editor".to_string()),
                },
                SubmissionNode::Not {
                    child: Box::new(SubmissionNode::Leaf {
                        left: Operand::Param {
                            name: "amount".to_string(),
                        },
                        op: Operator::Gt,
                        right: Operand::Static {
                            value: json!(1000),
                        },
                        failure_message: None,
                    }),
                    failure_message: None,
                },
            ],
        };

        let raw = serde_json::to_value(&node).expect("serialize");
        let back: SubmissionNode = serde_json::from_value(raw).expect("deserialize");
        assert_eq!(node, back);
    }
}

File: libs/ontology-kernel/models/submission_criteria.go
package models

import (
	"encoding/json"
	"errors"
	"fmt"
)

// Package-level note: this file ports
// `libs/ontology-kernel/src/models/submission_criteria.rs` (TASK C —
// Submission criteria AST). The Rust types use `#[serde(tag = "kind",
// rename_all = "snake_case")]` for `Operand` and `#[serde(tag =
// "type", rename_all = "snake_case")]` for `SubmissionNode`. Rust
// emits enum variants using the discriminant key + the variant
// fields side-by-side at the top level of the JSON object. Custom
// MarshalJSON/UnmarshalJSON below preserve byte-for-byte output.

// Operand mirrors `enum Operand` with `#[serde(tag = "kind",
// rename_all = "snake_case")]`.
type Operand struct {
	Kind     OperandKind     `json:"-"`
	Param    *OperandParam    `json:"-"`
	ParamProp *OperandParamProperty `json:"-"`
	User     *OperandCurrentUser `json:"-"`
	Static   *OperandStatic     `json:"-"`
}

type OperandKind string

const (
	OperandKindParam         OperandKind = "param"
	OperandKindParamProperty OperandKind = "param_property"
	OperandKindCurrentUser   OperandKind = "current_user"
	OperandKindStatic        OperandKind = "static"
)

// OperandParam: { "kind": "param", "name": ... }
type OperandParam struct {
	Name string `json:"name"`
}

// OperandParamProperty: { "kind": "param_property", "param": ...,
// "property": ... }
type OperandParamProperty struct {
	Param    string `json:"param"`
	Property string `json:"property"`
}

// OperandCurrentUser: { "kind": "current_user", "attribute": ... }
type OperandCurrentUser struct {
	Attribute UserAttr `json:"attribute"`
}

// OperandStatic: { "kind": "static", "value": ... }
type OperandStatic struct {
	Value json.RawMessage `json:"value"`
}

// MarshalJSON emits the operand with the `kind` discriminant inlined
// at the top level — same shape Rust serde produces.
func (o Operand) MarshalJSON() ([]byte, error) {
	switch o.Kind {
	case OperandKindParam:
		if o.Param == nil {
			return nil, errors.New("Operand.Param is nil for kind=param")
		}
		return json.Marshal(struct {
			Kind OperandKind `json:"kind"`
			*OperandParam
		}{Kind: o.Kind, OperandParam: o.Param})
	case OperandKindParamProperty:
		if o.ParamProp == nil {
			return nil, errors.New("Operand.ParamProp is nil for kind=param_property")
		}
		return json.Marshal(struct {
			Kind OperandKind `json:"kind"`
			*OperandParamProperty
		}{Kind: o.Kind, OperandParamProperty: o.ParamProp})
	case OperandKindCurrentUser:
		if o.User == nil {
			return nil, errors.New("Operand.User is nil for kind=current_user")
		}
		return json.Marshal(struct {
			Kind OperandKind `json:"kind"`
			*OperandCurrentUser
		}{Kind: o.Kind, OperandCurrentUser: o.User})
	case OperandKindStatic:
		if o.Static == nil {
			return nil, errors.New("Operand.Static is nil for kind=static")
		}
		return json.Marshal(struct {
			Kind  OperandKind     `json:"kind"`
			Value json.RawMessage `json:"value"`
		}{Kind: o.Kind, Value: o.Static.Value})
	default:
		return nil, fmt.Errorf("unknown operand kind %q", o.Kind)
	}
}

// UnmarshalJSON dispatches on the `kind` discriminant. Mirrors Rust
// serde with `tag = "kind"`.
func (o *Operand) UnmarshalJSON(b []byte) error {
	var head struct {
		Kind OperandKind `json:"kind"`
	}
	if err := json.Unmarshal(b, &head); err != nil {
		return err
	}
	o.Kind = head.Kind
	switch head.Kind {
	case OperandKindParam:
		var v OperandParam
		if err := json.Unmarshal(b, &v); err != nil {
			return err
		}
		o.Param = &v
	case OperandKindParamProperty:
		var v OperandParamProperty
		if err := json.Unmarshal(b, &v); err != nil {
			return err
		}
		o.ParamProp = &v
	case OperandKindCurrentUser:
		var v OperandCurrentUser
		if err := json.Unmarshal(b, &v); err != nil {
			return err
		}
		o.User = &v
	case OperandKindStatic:
		var v OperandStatic
		if err := json.Unmarshal(b, &v); err != nil {
			return err
		}
		o.Static = &v
	default:
		return fmt.Errorf("unknown operand kind %q", head.Kind)
	}
	return nil
}

// UserAttr mirrors `enum UserAttr` (`rename_all = "snake_case"`).
type UserAttr string

const (
	UserAttrUserID         UserAttr = "user_id"
	UserAttrEmail          UserAttr = "email"
	UserAttrOrganizationID UserAttr = "organization_id"
	UserAttrRoles          UserAttr = "roles"
	UserAttrPermissions    UserAttr = "permissions"
	UserAttrAuthMethods    UserAttr = "auth_methods"
)

// Operator mirrors `enum Operator` (`rename_all = "snake_case"`). 14
// variants — keep this list verbatim to match the Rust definition.
type Operator string

const (
	OperatorIs           Operator = "is"
	OperatorIsNot        Operator = "is_not"
	OperatorMatches      Operator = "matches"
	OperatorLt           Operator = "lt"
	OperatorLte          Operator = "lte"
	OperatorGt           Operator = "gt"
	OperatorGte          Operator = "gte"
	OperatorIncludes     Operator = "includes"
	OperatorIncludesAny  Operator = "includes_any"
	OperatorIsIncludedIn Operator = "is_included_in"
	OperatorEachIs       Operator = "each_is"
	OperatorEachIsNot    Operator = "each_is_not"
	OperatorIsEmpty      Operator = "is_empty"
	OperatorIsNotEmpty   Operator = "is_not_empty"
)

// SubmissionNodeType mirrors `enum SubmissionNode` discriminant `type`.
type SubmissionNodeType string

const (
	SubmissionNodeTypeLeaf SubmissionNodeType = "leaf"
	SubmissionNodeTypeAll  SubmissionNodeType = "all"
	SubmissionNodeTypeAny  SubmissionNodeType = "any"
	SubmissionNodeTypeNot  SubmissionNodeType = "not"
)

// SubmissionNode mirrors `enum SubmissionNode` with `#[serde(tag =
// "type", rename_all = "snake_case")]`. The four variants and their
// fields are flattened with `type` at the top level, matching Rust
// serde output.
//
// All four variants carry an optional `failure_message` with
// `#[serde(default, skip_serializing_if = "Option::is_none")]`.
type SubmissionNode struct {
	Type           SubmissionNodeType `json:"-"`
	Left           *Operand           `json:"-"`
	Op             Operator           `json:"-"`
	Right          *Operand           `json:"-"`
	Children       []SubmissionNode   `json:"-"`
	Child          *SubmissionNode    `json:"-"`
	FailureMessage *string            `json:"-"`
}

// NewLeaf mirrors `fn leaf(left, op, right)`.
func NewLeaf(left Operand, op Operator, right Operand) SubmissionNode {
	return SubmissionNode{
		Type:  SubmissionNodeTypeLeaf,
		Left:  &left,
		Op:    op,
		Right: &right,
	}
}

// FailureMessageStr mirrors `fn failure_message(&self) -> Option<&str>`.
func (n SubmissionNode) FailureMessageStr() *string { return n.FailureMessage }

// MarshalJSON emits each variant in the Rust serde shape.
func (n SubmissionNode) MarshalJSON() ([]byte, error) {
	switch n.Type {
	case SubmissionNodeTypeLeaf:
		return json.Marshal(struct {
			Type           SubmissionNodeType `json:"type"`
			Left           Operand            `json:"left"`
			Op             Operator           `json:"op"`
			Right          Operand            `json:"right"`
			FailureMessage *string            `json:"failure_message,omitempty"`
		}{
			Type:           n.Type,
			Left:           *n.Left,
			Op:             n.Op,
			Right:          *n.Right,
			FailureMessage: n.FailureMessage,
		})
	case SubmissionNodeTypeAll, SubmissionNodeTypeAny:
		return json.Marshal(struct {
			Type           SubmissionNodeType `json:"type"`
			Children       []SubmissionNode   `json:"children"`
			FailureMessage *string            `json:"failure_message,omitempty"`
		}{
			Type:           n.Type,
			Children:       n.Children,
			FailureMessage: n.FailureMessage,
		})
	case SubmissionNodeTypeNot:
		return json.Marshal(struct {
			Type           SubmissionNodeType `json:"type"`
			Child          *SubmissionNode    `json:"child"`
			FailureMessage *string            `json:"failure_message,omitempty"`
		}{
			Type:           n.Type,
			Child:          n.Child,
			FailureMessage: n.FailureMessage,
		})
	default:
		return nil, fmt.Errorf("unknown submission node type %q", n.Type)
	}
}

// UnmarshalJSON dispatches on the `type` discriminant.
func (n *SubmissionNode) UnmarshalJSON(b []byte) error {
	var head struct {
		Type SubmissionNodeType `json:"type"`
	}
	if err := json.Unmarshal(b, &head); err != nil {
		return err
	}
	n.Type = head.Type
	switch head.Type {
	case SubmissionNodeTypeLeaf:
		var v struct {
			Left           Operand  `json:"left"`
			Op             Operator `json:"op"`
			Right          Operand  `json:"right"`
			FailureMessage *string  `json:"failure_message"`
		}
		if err := json.Unmarshal(b, &v); err != nil {
			return err
		}
		n.Left = &v.Left
		n.Op = v.Op
		n.Right = &v.Right
		n.FailureMessage = v.FailureMessage
	case SubmissionNodeTypeAll, SubmissionNodeTypeAny:
		var v struct {
			Children       []SubmissionNode `json:"children"`
			FailureMessage *string          `json:"failure_message"`
		}
		if err := json.Unmarshal(b, &v); err != nil {
			return err
		}
		n.Children = v.Children
		n.FailureMessage = v.FailureMessage
	case SubmissionNodeTypeNot:
		var v struct {
			Child          *SubmissionNode `json:"child"`
			FailureMessage *string         `json:"failure_message"`
		}
		if err := json.Unmarshal(b, &v); err != nil {
			return err
		}
		n.Child = v.Child
		n.FailureMessage = v.FailureMessage
	default:
		return fmt.Errorf("unknown submission node type %q", head.Type)
	}
	return nil
}
